### PR TITLE
Better brackets completation

### LINKED ISSUE #31: Brackets completion should also un-complete
If I type an opening bracket/paren/quote, then backspace it, the Brackets Completion plugin should also removed its added closing bracket/paren/quote. Right now it just leaves it.

Atom and Sublime have great behavior here if we need to investigate edge cases.

Launchpad Details: [#LP1547635](https://bugs.launchpad.net/bugs/1547635) Cassidy James Blaede - 2016-02-19 18:55:34 +0000

### PR DESCRIPTION
**Pull request by snowparrot**
intended to fix #31 but deletes two or more letters. In contrast a test app I wrote works as intended.
So as I asked in Slack may a dev look at it and give me a clue?
[bracketsTest.tar.gz](https://github.com/elementary/scratch/files/1246168/bracketsTest.tar.gz)

### CODE REVIEW

File: plugins/brackets-completion/brackets-completion.vala [hunk 1]
```diff
@@ -23,10 +23,14 @@ public const string DESCRIPTION = _("Complete brackets while typing");
 
 public class Scratch.Plugins.BracketsCompletion : Peas.ExtensionBase,  Peas.Activatable {
     Gee.HashMap<string, string> brackets;
+    // saves, when left bracket is deleted, if right bracket will also be deleted 
+    Gee.HashMap<string, bool> bracketsDALD;
```
Comment (matzipan) [inline plugins/brackets-completion/brackets-completion.vala]: Please refer to the code style guide for how to name variables: https://elementary.io/docs/code/reference#variable-names-class-names-function-names

File: plugins/brackets-completion/brackets-completion.vala [hunk 2]
```diff
@@ -23,10 +23,14 @@ public const string DESCRIPTION = _("Complete brackets while typing");
...
 
     Gee.TreeSet<Gtk.TextBuffer> buffers;
     Gtk.TextBuffer current_buffer;
     string last_inserted;
+    
+    bool AttentionBracket;
```
Comment (matzipan) [inline plugins/brackets-completion/brackets-completion.vala]: Same here: https://elementary.io/docs/code/reference#variable-names-class-names-function-names

File: plugins/brackets-completion/brackets-completion.vala [hunk 3]
```diff
@@ -48,14 +53,59 @@ public class Scratch.Plugins.BracketsCompletion : Peas.ExtensionBase,  Peas.Acti
         this.brackets.set ("‘", "‘");
         this.brackets.set ("'", "'");
         this.brackets.set ("\"", "\"");
-
+
```
Comment (matzipan) [inline plugins/brackets-completion/brackets-completion.vala]: Remove one line of whitespace.

File: plugins/brackets-completion/brackets-completion.vala [hunk 4]
```diff
@@ -48,14 +53,59 @@ public class Scratch.Plugins.BracketsCompletion : Peas.ExtensionBase,  Peas.Acti
...
+            
+            doc.source_view.move_cursor.connect ((a, b, c) => {
+                // if cursor moves around code doesnt care about brackets anymore
+                AttentionBracket = false;
+            });
+            
+            doc.source_view.backspace.connect (() => {
+                if (AttentionBracket) {
+                    
+                    Gtk.TextIter leftiter;
```
Comment (matzipan) [inline plugins/brackets-completion/brackets-completion.vala]: Again, naming convention: https://elementary.io/docs/code/reference#variable-names-class-names-function-names

File: plugins/brackets-completion/brackets-completion.vala [hunk 5]
```diff
@@ -48,14 +53,59 @@ public class Scratch.Plugins.BracketsCompletion : Peas.ExtensionBase,  Peas.Acti
...
+                    buf.get_iter_at_mark(out leftiter, buf.get_insert());
+                    
+                    var rightiter = leftiter;
```
Comment (matzipan) [inline plugins/brackets-completion/brackets-completion.vala]: Here too: https://elementary.io/docs/code/reference#variable-names-class-names-function-names

File: plugins/brackets-completion/brackets-completion.vala [hunk 6]
```diff
@@ -48,14 +53,59 @@ public class Scratch.Plugins.BracketsCompletion : Peas.ExtensionBase,  Peas.Acti
...
+                    
+                    //leftiter.backward_cursor_positions (1);
```
Comment (matzipan) [inline plugins/brackets-completion/brackets-completion.vala]: Do not leave comments in merged code

File: plugins/brackets-completion/brackets-completion.vala [hunk 7]
```diff
@@ -48,14 +53,59 @@ public class Scratch.Plugins.BracketsCompletion : Peas.ExtensionBase,  Peas.Acti
...
+                     rightiter.forward_cursor_positions (1);
+                     buf.@delete(ref leftiter, ref rightiter);       
+                
+                    // https://valadoc.org/gtk+-3.0/Gtk.TextBuffer.get_insert.html hier bekommt TextMark
```
Comment (matzipan) [inline plugins/brackets-completion/brackets-completion.vala]: Do not leave comments in merged code

File: plugins/brackets-completion/brackets-completion.vala [hunk 8]
```diff
@@ -23,10 +23,14 @@ public const string DESCRIPTION = _("Complete brackets while typing");
 
 public class Scratch.Plugins.BracketsCompletion : Peas.ExtensionBase,  Peas.Activatable {
     Gee.HashMap<string, string> brackets;
+    // saves, when left bracket is deleted, if right bracket will also be deleted
```
Comment (matzipan) [inline plugins/brackets-completion/brackets-completion.vala]: Please rephrase this, as it sounds a bit confusing.

**Comment by snowparrot** [comment]
@matzipan Thank you for your help though  I admit it wasn't my intention to have clear code base now. :P Do you have any clue about functionality?

**Comment by matzipan** [comment]
Ah, I misunderstood your initial comment then, sorry.

I don't see what the problem is with the branch? It appears to behave as expected.

**Comment by matzipan** [comment]
However, I think in the case of quotes `"`, when the user wants only one of them, it pressed the _delete_ key, which deletes the autocompleted quote, not the typed one.

**Comment by matzipan** [comment]
Regarding the last comment, this is the sort of behaviour I see in Atom for both `{` and `"`.

**Comment by snowparrot** [comment]
@matzipan now it works :)
I guess it is a question of user experience. @cassidyjames As I and probably matzipan don't decide UX I would like to see your decision which behaviour should be implented.

**Comment by matzipan** [comment]
As per original issue description, Atom and Sublime functionality are a good reference.

**Comment by snowparrot** [comment]
@matzipan Done! acts like Atom.

**Comment by snowparrot** [comment]
#203 includes my pull request.

**Pull request closed without merging**